Source/devices/oni/rhythm-api/rhd2000ONIboard.cpp

```diff
@@ -1204,43 +1206,80 @@ bool Rhd2000ONIBoard::isI2cCapable(oni_dev_idx_t device)
     return val > 0;
 }
 
-int Rhd2000ONIBoard::readByte(uint32_t address, oni_reg_addr_t i2cRawAddress, oni_reg_val_t* value, bool sixteenBitAddress)
+int Rhd2000ONIBoard::readByte (oni_dev_idx_t device, uint32_t address, oni_reg_val_t* value)
+{
+    if (! ctx)
+        return 0;
+
+    int result = oni_read_reg (ctx, device, address, value);
+
+    return result;
+}
+
+int Rhd2000ONIBoard::readEepromByte (oni_dev_idx_t i2cRawAddress, uint32_t address, uint8_t& byte)
 {
-    uint32_t i2cAddress = 0x50; 
+    uint32_t i2cAddress = 0x50;
 
     uint32_t registerAddress = (address << 7) | (i2cAddress & 0x7F);
-    registerAddress |= sixteenBitAddress ? 0x80000000 : 0;
+    registerAddress |= 0x80000000;
 
-    int result = oni_read_reg (ctx, i2cRawAddress, registerAddress, value);
+    oni_reg_val_t value;
+
+    int result = readByte (i2cRawAddress, registerAddress, &value);
+
+    if (result == 0)
+        byte = (uint8_t) (value & 0xFF);
+    else
+        byte = 0;
 
     return result;
 }
 
-uint32_t Rhd2000ONIBoard::getDeviceIdOnEeprom (oni_reg_addr_t i2cRawAddress)
+uint32_t Rhd2000ONIBoard::getDeviceIdOnEeprom (const uint32_t port)
 {
-    if (i2cRawAddress != DEVICE_I2C_RAW_A && i2cRawAddress != DEVICE_I2C_RAW_B && i2cRawAddress != DEVICE_I2C_RAW_C && i2cRawAddress != DEVICE_I2C_RAW_D)
+    if (!ctx || port > 3)
         return 0;
 
+    oni_dev_idx_t i2cRawAddress = DEVICE_I2C_RAW_A + port * 2;
+
+    // NB: Confirm EEPROM first four bytes contains OESH
+    const char identifier[] = "OESH";
+
+    for (unsigned int i = 0; i < strlen(identifier); i += 1)
+    {
+        uint8_t val;
+        int res;
+        res = readEepromByte (i2cRawAddress, i, val);
+
+        if (res != 0 || val != identifier[i])
+            return 0;
+    }
+
     const uint32_t deviceIdStartAddress = 7;
     uint32_t data = 0;
 
     for (unsigned int i = 0; i < sizeof (uint32_t); i++)
     {
-        oni_reg_val_t val;
+        uint8_t val;
         int res;
-        res = readByte (deviceIdStartAddress + i, i2cRawAddress, &val, true);
-        data += (val & 0xFF) << (8 * i);
+        res = readEepromByte (i2cRawAddress, deviceIdStartAddress + i, val);
+
+        if (res != 0)
+            return 0;
+
+        data += val << (8 * i);
     }
 
     return data;
 }
 
-bool Rhd2000ONIBoard::isBnoConnected (oni_dev_idx_t device)
+bool Rhd2000ONIBoard::isBnoConnected (const uint32_t port)
 {
-    if (! ctx || (device != DEVICE_BNO_A && device != DEVICE_BNO_B && device != DEVICE_BNO_C && device != DEVICE_BNO_D))
+    if (! ctx || port > 3)
         return false;
 
     oni_reg_val_t val = 2; // Value == 2 means there is a BNO scan in progress
+    oni_dev_idx_t device = DEVICE_I2C_RAW_A + port * 2;
```

Comment:
This is using the wrong address. It should use the bno address as base, since you are talking with the bno here, not the i2c raw.